cmd/git/main.go

```diff
@@ -0,0 +1,356 @@
+				return err
+			}
+
+			var sshCmd = []string{"ssh", "-i", sshPrivateKeyFile.Name()}
+
+			var knownHostsFile = filepath.Join(flagValues.secretPath, "known_hosts")
+			if hasFile(knownHostsFile) {
+				sshCmd = append(sshCmd,
+					"-o", "GlobalKnownHostsFile=/dev/null",
+					"-o", fmt.Sprintf("UserKnownHostsFile=%s", knownHostsFile),
+				)
+			} else {
+				sshCmd = append(sshCmd,
+					"-o", "StrictHostKeyChecking=accept-new",
+				)
+			}
+
+			args = append(args,
+				"--config",
+				fmt.Sprintf(`core.sshCommand=%s`, strings.Join(sshCmd, " ")),
+			)
+
+		case typeUsernamePassword:
+			repoURL, err := url.Parse(flagValues.url)
+			if err != nil {
+				return err
+			}
+
+			username, err := ioutil.ReadFile(filepath.Join(flagValues.secretPath, "username"))
+			if err != nil {
+				return err
+			}
+
+			password, err := ioutil.ReadFile(filepath.Join(flagValues.secretPath, "password"))
+			if err != nil {
+				return err
+			}
+
+			repoURL.User = url.UserPassword(string(username), string(password))
+
+			credHelperFile, err := ioutil.TempFile(os.TempDir(), "cred-helper-file")
+			if err != nil {
+				return err
+			}
+
+			defer os.Remove(credHelperFile.Name())
+
+			if err := ioutil.WriteFile(credHelperFile.Name(), []byte(repoURL.String()), 0400); err != nil {
+				return err
+			}
+
+			if _, err := git(ctx, "config", "--global", "credential.helper", fmt.Sprintf("store --file %s", credHelperFile.Name())); err != nil {
+				return err
+			}
+		}
+	}
+
+	if _, err := git(ctx, append(args, "--", flagValues.url, flagValues.target)...); err != nil {
+		return err
+	}
+
+	if commitSha != "" {
+		_, err := git(ctx, "-C", flagValues.target, "checkout", commitSha)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err := git(ctx, "-C", flagValues.target, "submodule", "update", "--init", "--recursive")
+	return err
+}
+
+func git(ctx context.Context, args ...string) (string, error) {
+	cmd := exec.CommandContext(ctx, "git", args...)
+	ctxlog.Debug(ctx, cmd.String())
+
+	out, err := cmd.CombinedOutput()
+
+	var output string
+	if out != nil {
+		output = strings.TrimRight(string(out), "\n")
+	}
+
+	if err != nil {
+		// In case the command fails, it is very likely to be an ExitCode error
+		// which contains the exit code of the command. Create a custom exit
+		// error where the command output is placed into the error to be more
+		// readable to the end-user.
+		switch terr := err.(type) {
+		case *exec.ExitError:
+			err = &ExitError{
+				Code:    terr.ExitCode(),
+				Message: output,
+				Cause:   err,
+			}
+		}
+
+		ctxlog.Error(ctx, err, "git command failed",
+			"command", cmd.String(),
+			"output", output,
+		)
+	}
+
+	return output, err
+}
+
+func hasFile(elem ...string) bool {
+	_, err := os.Stat(filepath.Join(elem...))
+	return !os.IsNotExist(err)
+}
+
+func checkCredentials() (credentialType, error) {
+	hasPrivateKey := hasFile(flagValues.secretPath, "ssh-privatekey")
```

Comment:
can we document the need of the private key to be under a file called `ssh-privatekey` ?

Reply:
I can put a comment that the name comes from the `kubernetes.io/ssh-auth` secret type.

Reply:
yes, that would help.